Flexvolume driver

vefimova opened the pull request:
Added flexvolume libvirt driver to support attaching rbd block devices to VMs.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mirantis/virtlet/239)
<!-- Reviewable:end -->

vefimova commented on docs/volumes.md:
```diff
@@ -52,4 +52,117 @@ Additional volume disks are exposed in the alphabet order starting from b, so vo
 ## Persistent Storage
-TODO
+## Flexvolume libvirt driver
+**NOTE: Currently, blocked by qemu connectivity issue, see details in !!!**
```
Oh, forgot about issue for connectivity, wanted to add link here to it, will create and update.

nhlfr commented:
Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[image_skel/etc/libvirt/libvirtd.conf, line 5 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2Yb-STJYyhPiDYb-5:-Kh2Yb-STJYyhPiDYb-6:b-edh4c) ([raw file](https://github.com/mirantis/virtlet/blob/3d82626c5ffb4146f4c64f7a03a3b50e3088d77d/image_skel/etc/libvirt/libvirtd.conf#L5)):*
> ```
> ca_file = ""
> log_level = 1
> log_outputs = "1:file:/tmp/libvirt.log"
> ```

What's the reason for that change? If I understand correctly, this will make the libvirt logs unavailable under `kubectl logs` and instead user will need to enter the container and read the file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

vefimova commented:
Review status: 0 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[image_skel/etc/libvirt/libvirtd.conf, line 5 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2Yb-STJYyhPiDYb-5:-Kh2_QpTX_mchX4k9PQA:b-fzxe64) ([raw file](https://github.com/mirantis/virtlet/blob/3d82626c5ffb4146f4c64f7a03a3b50e3088d77d/image_skel/etc/libvirt/libvirtd.conf#L5)):*
<details><summary><i>Previously, nhlfr (Michal Rostecki) wrote…</i></summary><blockquote>

What's the reason for that change? If I understand correctly, this will make the libvirt logs unavailable under `kubectl logs` and instead user will need to enter the container and read the file.
</blockquote></details>

yep thx, didn't want to add it, changed only for debugging, fixed. BTW "x:stderr" is invalid record.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

ivan4th commented:
Review status: 0 of 14 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/libvirttools/storage_pool.go, line 285 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh3-dK7nYMI8nUhSPyg:-Kh3-dK8dTagF6lD6Cmi:b6hfc6j) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/storage_pool.go#L285)):*
> ```Go
> 
> 	for ind, virtletVol := range virtletVols {
> 		if ind == len(diskLetter) {
> ```

so the contents of diskLetter is not used here, any reason to pass it to this func?

---

*[pkg/libvirttools/virtualization.go, line 321 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2zwSvA3peGtkUCLkR:-Kh2zwSvA3peGtkUCLkS:b-70cz2j) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/virtualization.go#L321)):*
> ```Go
> func addDiskToDomainDefinition(domain *Domain, volXML string) error {
> 	disk := &Disk{}
> 	err := xml.Unmarshal([]byte(volXML), disk)
> ```

can write this inside if

---

*[pkg/libvirttools/virtualization.go, line 338 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2yygYhJA7O51oDvew:-Kh2yygYhJA7O51oDvex:bq3j0ah) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/virtualization.go#L338)):*
> ```Go
> 
> func (v *VirtualizationTool) addAttachedVolumesXML(podID string, uuid string, virtletVolsDesc string, domXML string) (string, error) {
> 	diskLetter := strings.Split("bcdefghijklmnopqrstu", "")
> ```

It looks like `diskLetter` (`diskLetters` maybe?) is always the same? Any reason to pass it as an argument?

---

*[pkg/libvirttools/virtualization.go, line 380 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2zku0ObvlnaH7upY9:-Kh2zku1spCK5xJw9mYp:b-70cz2j) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/virtualization.go#L380)):*
> ```Go
> 
> 	for _, volXML := range volumesXML {
> 		err = addDiskToDomainDefinition(domain, volXML)
> ```

can write this inside if

---

*[tests/e2e/run_ceph.sh, line 21 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2b9zD7sBKny4RP20h:-Kh2b9zD7sBKny4RP20i:bjkbuxj) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/tests/e2e/run_ceph.sh#L21)):*
> ```Shell
> set -o errtrace
> 
> SCRIPT_DIR=$1
> ```

`"${1}"` woud be better probably

---

*[tests/e2e/run_ceph.sh, line 31 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2_qtzQIGUD6s_gSh9:-Kh2_qtzQIGUD6s_gShA:bv4pond) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/tests/e2e/run_ceph.sh#L31)):*
> ```Shell
> # Check cluster is running
> set +e
> if ! docker exec ${container_name} ceph -s 2> /dev/null 1> /dev/null; then
> ```

There may be a race, because there's no guarantee CEPH cluster is running at this point. (unless `ceph -s` waits for it... which I doubt)

---

*[tests/e2e/run_ceph.sh, line 39 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2bGgeF9lW3yN22Sbl:-Kh2bGgeF9lW3yN22Sbm:bb6rc9x) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/tests/e2e/run_ceph.sh#L39)):*
> ```Shell
> # Adjust ceph configs
> docker exec ${container_name} /bin/bash -c 'echo "rbd default features = 1
> rbd default format = 2" >> /etc/ceph/ceph.conf'
> ```

Maybe use `;` instead of newline?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

jellonek commented:
Impressive :)

---

Reviewed 13 of 15 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[cmd/flexvolume/libvirt_driver.go, line 1 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh31vy5t1Sm36m7yfbi:-Kh31vy5t1Sm36m7yfbj:bpmjmlg) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/cmd/flexvolume/libvirt_driver.go#L1)):*
> ```Go
> /*
> ```

IMO executable (and file with main routine) should have a name of directory. Once it's named `flexvolume`, once `libvirt_driver`. Name should be the same.

---

*[cmd/flexvolume/libvirt_driver.go, line 51 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh31mP68VIVdJHnvqpP:-Kh31mP68VIVdJHnvqpQ:b-ns6q3q) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/cmd/flexvolume/libvirt_driver.go#L51)):*
> ```Go
> }
> 
> func writeToFile(filePath, stringToWrite string) error {
> ```

Unnecessary func which needs coverage. Instead better would be to use https://golang.org/pkg/io/ioutil/#WriteFile from stdlib.

---

*[cmd/flexvolume/libvirt_driver.go, line 152 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh32Q6ri5klyqdIeEqU:-Kh32Q6ri5klyqdIeEqV:bsny1f0) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/cmd/flexvolume/libvirt_driver.go#L152)):*
> ```Go
> 
> func main() {
> 	switch os.Args[1] {
> ```

This can panic on no arguments.

---

*[cmd/flexvolume/libvirt_driver.go, line 156 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh32TCinCl6bTA9R197:-Kh32TCinCl6bTA9R198:b-yxwa0v) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/cmd/flexvolume/libvirt_driver.go#L156)):*
> ```Go
> 		Init()
> 	case "attach":
> 		Attach(os.Args[2])
> ```

This can panic on lack of required argument.

---

*[cmd/flexvolume/libvirt_driver.go, line 160 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh32X7FhTHUu5qt184t:-Kh32X7FhTHUu5qt184u:b-hbcorw) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/cmd/flexvolume/libvirt_driver.go#L160)):*
> ```Go
> 		Detach()
> 	case "mount":
> 		Mount(os.Args[2], os.Args[3], os.Args[4])
> ```

The same.

---

*[cmd/flexvolume/libvirt_driver.go, line 162 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh32_A1d99C8GvPTDfd:-Kh32_A1d99C8GvPTDfe:b-hbcorw) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/cmd/flexvolume/libvirt_driver.go#L162)):*
> ```Go
> 		Mount(os.Args[2], os.Args[3], os.Args[4])
> 	case "unmount":
> 		Unmount(os.Args[2])
> ```

The same.

---

*[deploy/virtlet-ds-dev.yaml, line 141 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh337c25wN8GubDEbks:-Kh337c25wN8GubDEbkt:b-807afv) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/deploy/virtlet-ds-dev.yaml#L141)):*
> ```YAML
>         - mountPath: /vmwrapper
>           name: vmwrapper-dev-bin
>         - mountPath: /libvirt_dirver
> ```

Typo.

---

*[docs/volumes.md, line 55 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh32qxf6tx2L0XkJtNK:-Kh32qxf6tx2L0XkJtNL:b2hx75i) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/docs/volumes.md#L55)):*
> ```Markdown
> 
> ## Persistent Storage
> ## Flexvolume libvirt driver
> ```

This should be probably started with ###.

---

*[docs/volumes.md, line 60 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh33tGUoceud1E5MSU4:-Kh33tGUoceud1E5MSU5:b98j2fo) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/docs/volumes.md#L60)):*
> ```Markdown
> FlexVolume libvirt driver for virtlet supports attaching to VM of Ceph RBD block devices from cluster with cephx auth enabled.
> 
> Basing on [FlexVolumes](https://github.com/kubernetes/community/blob/master/contributors/devel/flexvolume.md) for k8s implemeted defined api to have virtlet be alinged with a way of definition of  remote persistent volumes for pods inside pods(https://kubernetes.io/docs/concepts/storage/volumes/#flexvolume) as well as using [PV&PVC](https://kubernetes.io/docs/concepts/storage/persistent-volumes/)
> ```

... for k8s implement*s* defined api ...
Double space in paragraph in "of  remote". Missing space in "pods(...".

---

*[docs/volumes.md, line 115 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh34nYgycwiszY6VWcu:-Kh34nYhJBN3GPE14CW3:b-lk5e5t) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/docs/volumes.md#L115)):*
> ```Markdown
> AQDTwuVY8rA8HxAAthwOKaQPr0hRc7kCmR/9Qg==
> ```
> 4. Virtlet checks whether there're dirs with volume info under `/var/lib/kubelet/pods/<pod-id>/volumes/kubernetes.io~libvirt_driver`. If there're virtlet integrates disk.xml inside domain's definition and creates secret entity in libvirt for cephx auth basing on provided secret.xml and key.
> ```

s/there're/there are/

---

*[docs/volumes.md, line 159 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh350KRxwOWUP28ph_F:-Kh350KRxwOWUP28ph_G:btxozz9) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/docs/volumes.md#L159)):*
> ```Markdown
> ```
> 
> **NOTE: There's no support for _MountPoints_ specified for container, i.e. all defined volumes will be attached to VM.**
> ```

Attached where?
Provided to vm?

---

*[examples/cirros-vm-rbd-volume.yaml.tmpl, line 9 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh35AZ6tk7mKc4jaNKi:-Kh35AZ6tk7mKc4jaNKj:bow36kd) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/examples/cirros-vm-rbd-volume.yaml.tmpl#L9)):*
> ```
>     scheduler.alpha.kubernetes.io/affinity: >
>       {
>         \\"nodeAffinity\\": {
> ```

Those double `\\` are intentional?

---

*[pkg/libvirttools/storage_pool.go, line 285 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh3-dK7nYMI8nUhSPyg:-Kh36Dax-yzEhOqr5OaF:b-6qxw4) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/storage_pool.go#L285)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

so the contents of diskLetter is not used here, any reason to pass it to this func?
</blockquote></details>

Yeah, instead of passing `diskLetter` there is needed only it's length. Btw. also name of this variable seems to be incorrect, as it may contain more than single character.

---

*[pkg/libvirttools/virtualization.go, line 320 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh38dm6psYgIW8NYxAI:-Kh38dm6psYgIW8NYxAJ:bv486qm) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/virtualization.go#L320)):*
> ```Go
> 
> func addDiskToDomainDefinition(domain *Domain, volXML string) error {
> 	disk := &Disk{}
> ```

This should be imo:
```disk := Disk{}```
later used with `&` in `Unmarshal`, and without dereference in `append`.
```


---

*[pkg/libvirttools/virtualization.go, line 321 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2zwSvA3peGtkUCLkR:-Kh38KkK8iTniC9vh4ok:b-nzartd) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/virtualization.go#L321)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

can write this inside if
</blockquote></details>

Yeah, these lines should be combined.

---

*[pkg/libvirttools/virtualization.go, line 338 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2yygYhJA7O51oDvew:-Kh39GCw-JEP72rl7Dch:b-4azt2p) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/virtualization.go#L338)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

It looks like `diskLetter` (`diskLetters` maybe?) is always the same? Any reason to pass it as an argument?
</blockquote></details>

Not exactly.  Later, in call to `CreateVolumesToBeAttached` only part of this slice is passed.
IMO there is no need to pass it at all. Instead we could pass number of already allocated disks, so there would be also no need to pass length of whole slice, as it would be constant. Also then this slice can be declared as constant for whole package.

---

*[pkg/libvirttools/virtualization.go, line 343 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh39hkA2hE626yURWxa:-Kh39hkA2hE626yURWxb:b-8fx2u7) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/virtualization.go#L343)):*
> ```Go
> 	err := xml.Unmarshal([]byte(domXML), domain)
> 	if err != nil {
> 		return "", err
> ```

:+1: 

---

*[tests/e2e/run_ceph.sh, line 24 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh3C3AySVyMRxVYrILP:-Kh3C3AySVyMRxVYrILQ:bnbo0yy) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/tests/e2e/run_ceph.sh#L24)):*
> ```Shell
> # Get default gateway ip from kube-master node deployed by kubeadm-dind tool
> MON_IP=$(docker exec kube-master route | grep default | awk '{print $2}')
> CEPH_PUBLIC_NETWORK=${MON_IP}/16
> ```

Value should be probably enclosed in `""`. MON_IP should be always string without white chars, but to be sure, each use of `$VARIABLE` we should enclose in this way (so also in docker invocation below).

---

*[tests/e2e/run_ceph.sh, line 39 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2bGgeF9lW3yN22Sbl:-Kh3Clo3-Zpom0CT7C17:b-zemxj) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/tests/e2e/run_ceph.sh#L39)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

Maybe use `;` instead of newline?
</blockquote></details>

Yeah, 2x echo with redir (quite readable), or `echo -e "...\n..."` (a bit less readable, but single command)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239#-:-Kh3DN9MPlc6QP-6ztmc:bxb913l)*
<!-- Sent from Reviewable.io -->

vefimova commented:
Review status: all files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[cmd/flexvolume/libvirt_driver.go, line 1 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh31vy5t1Sm36m7yfbi:-Kh3PoMGgcYa5v1klFtj:b32wtd) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/cmd/flexvolume/libvirt_driver.go#L1)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

IMO executable (and file with main routine) should have a name of directory. Once it's named `flexvolume`, once `libvirt_driver`. Name should be the same.
</blockquote></details>

ok

---

*[cmd/flexvolume/libvirt_driver.go, line 51 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh31mP68VIVdJHnvqpP:-Kh3KuZlyDtLUY6IvbqQ:bqp23ws) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/cmd/flexvolume/libvirt_driver.go#L51)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Unnecessary func which needs coverage. Instead better would be to use https://golang.org/pkg/io/ioutil/#WriteFile from stdlib.
</blockquote></details>

thx, will remove it

---

*[cmd/flexvolume/libvirt_driver.go, line 152 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh32Q6ri5klyqdIeEqU:-Kh3Kf6xSoVO8TSkMfPV:blmwutu) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/cmd/flexvolume/libvirt_driver.go#L152)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

This can panic on no arguments.
</blockquote></details>

yeah, you're right, it's better to check this dirver api contract, as it could be changed unexpectedly and there's no versioning even.

---

*[cmd/flexvolume/libvirt_driver.go, line 156 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh32TCinCl6bTA9R197:-Kh3KzTqnUs9712mnlzo:b32wtd) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/cmd/flexvolume/libvirt_driver.go#L156)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

This can panic on lack of required argument.
</blockquote></details>

ok

---

*[cmd/flexvolume/libvirt_driver.go, line 160 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh32X7FhTHUu5qt184t:-Kh3L0BC8LmuuSTvA5Ue:b32wtd) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/cmd/flexvolume/libvirt_driver.go#L160)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

The same.
</blockquote></details>

ok

---

*[cmd/flexvolume/libvirt_driver.go, line 162 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh32_A1d99C8GvPTDfd:-Kh3L14k6yKUO9PQBlmb:b32wtd) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/cmd/flexvolume/libvirt_driver.go#L162)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

The same.
</blockquote></details>

ok

---

*[deploy/virtlet-ds-dev.yaml, line 141 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh337c25wN8GubDEbks:-Kh3LNO49w5tSGLYg-EN:b-t0dj1o) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/deploy/virtlet-ds-dev.yaml#L141)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Typo.
</blockquote></details>

thx, will fix. hah, even after you pointing it out, had some time looking, where... :)

---

*[docs/volumes.md, line 55 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh32qxf6tx2L0XkJtNK:-Kh3LcbMNqXis6xU4Afa:bgvs6h8) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/docs/volumes.md#L55)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

This should be probably started with ###.
</blockquote></details>

will fix

---

*[docs/volumes.md, line 60 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh33tGUoceud1E5MSU4:-Kh3LspKwjzOAnmHEv45:b32wtd) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/docs/volumes.md#L60)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

... for k8s implement*s* defined api ...
Double space in paragraph in "of  remote". Missing space in "pods(...".
</blockquote></details>

ok

---

*[docs/volumes.md, line 115 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh34nYgycwiszY6VWcu:-Kh3M-uSuHl0NTHaYO-k:b32wtd) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/docs/volumes.md#L115)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

s/there're/there are/
</blockquote></details>

ok

---

*[docs/volumes.md, line 159 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh350KRxwOWUP28ph_F:-Kh3OVbGWIC79Jdn8jbh:b-a24otz) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/docs/volumes.md#L159)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Attached where?
Provided to vm?
</blockquote></details>

not quite understand what is wrong here? Here I wanted to underline that unlike to containers, for VM-pods you don't need any additional settings inside container spec to have it attached to VM. Probabaly I shouldn't mention MountPoint, and maybe just leeave only this: "NOTE: All defined volumes will be attached to VM, no additional settings needed inside container spec." ?

---

*[examples/cirros-vm-rbd-volume.yaml.tmpl, line 9 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh35AZ6tk7mKc4jaNKi:-Kh3MS5NSFS5DaD9xsLd:b-6dxp8c) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/examples/cirros-vm-rbd-volume.yaml.tmpl#L9)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Those double `\\` are intentional?
</blockquote></details>

yep, as using eval to substitute vars in template, which will eat the quotes otherwise.

---

*[pkg/libvirttools/storage_pool.go, line 285 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh3-dK7nYMI8nUhSPyg:-Kh3MrXlGq0dXE3280Kb:ba1x7qw) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/storage_pool.go#L285)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Yeah, instead of passing `diskLetter` there is needed only it's length. Btw. also name of this variable seems to be incorrect, as it may contain more than single character.
</blockquote></details>

thx, will check re passing.
well, it's array of signle lenght strings :) add "Arr" to name?

---

*[pkg/libvirttools/virtualization.go, line 320 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh38dm6psYgIW8NYxAI:-Kh3Q30uDHhZA7syiH2l:bntyi3w) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/virtualization.go#L320)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

This should be imo:
```disk := Disk{}```
later used with `&` in `Unmarshal`, and without dereference in `append`.
</blockquote></details>

yep, thx, will fix

---

*[pkg/libvirttools/virtualization.go, line 321 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2zwSvA3peGtkUCLkR:-Kh3NLbNN7segK4oV97G:b32wtd) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/virtualization.go#L321)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Yeah, these lines should be combined.
</blockquote></details>

ok

---

*[pkg/libvirttools/virtualization.go, line 338 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2yygYhJA7O51oDvew:-Kh3QQ47JxbSo7y-V6LU:b2ni9ah) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/virtualization.go#L338)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Not exactly.  Later, in call to `CreateVolumesToBeAttached` only part of this slice is passed.
IMO there is no need to pass it at all. Instead we could pass number of already allocated disks, so there would be also no need to pass length of whole slice, as it would be constant. Also then this slice can be declared as constant for whole package.
</blockquote></details>

ok

---

*[pkg/libvirttools/virtualization.go, line 380 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2zku0ObvlnaH7upY9:-Kh3Nech10xYWCQshdKX:b32wtd) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/virtualization.go#L380)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

can write this inside if
</blockquote></details>

ok

---

*[tests/e2e/run_ceph.sh, line 21 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2b9zD7sBKny4RP20h:-Kh3Nfkpe_CcvttWAjHs:b32wtd) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/tests/e2e/run_ceph.sh#L21)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

`"${1}"` woud be better probably
</blockquote></details>

ok

---

*[tests/e2e/run_ceph.sh, line 24 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh3C3AySVyMRxVYrILP:-Kh3RQoCjmkW6mK-n5cT:b32wtd) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/tests/e2e/run_ceph.sh#L24)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Value should be probably enclosed in `""`. MON_IP should be always string without white chars, but to be sure, each use of `$VARIABLE` we should enclose in this way (so also in docker invocation below).
</blockquote></details>

ok

---

*[tests/e2e/run_ceph.sh, line 31 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2_qtzQIGUD6s_gSh9:-Kh3NlYES2JBP7arzSLp:b-ee9ohl) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/tests/e2e/run_ceph.sh#L31)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

There may be a race, because there's no guarantee CEPH cluster is running at this point. (unless `ceph -s` waits for it... which I doubt)
</blockquote></details>

yeah, thx, good catch, will add wait in loop here.

---

*[tests/e2e/run_ceph.sh, line 39 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2bGgeF9lW3yN22Sbl:-Kh3O-SaUoSGm8pua-Yd:b32wtd) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/tests/e2e/run_ceph.sh#L39)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Yeah, 2x echo with redir (quite readable), or `echo -e "...\n..."` (a bit less readable, but single command)?
</blockquote></details>

ok

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

vefimova commented:
Fixed comments.

---

Review status: 2 of 14 files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239#-:-KhDGZbb5GkiuV4vnXsL:b-ge3f07)*
<!-- Sent from Reviewable.io -->

ivan4th commented:
Mostly LGTM except for a some small nits

---

Review status: 2 of 14 files reviewed at latest revision, 14 unresolved discussions.

---

*[test.sh, line 14 at r3](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhMzvK7lQ73eA44OjMU:-KhMzvK7lQ73eA44OjMV:bu2ekya) ([raw file](https://github.com/mirantis/virtlet/blob/963f4c07801fb2672488f11f3460aba601c5e11f/test.sh#L14)):*
> ```Shell
> 
> cd "$(dirname "$(readlinkf "${BASH_SOURCE}")")"
> SCRIPT_DIR="$(cd $(dirname "$(readlinkf "${BASH_SOURCE}")"); pwd)"
> ```

hmm maybe just `cd "${SCRIPT_DIR}"` after `SCRIPT_DIR=...` ?

---

*[cmd/flexvolume_driver/flexvolume_driver.go, line 148 at r3](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhN-DaNqJfXg92UxcJI:-KhN-DaNqJfXg92UxcJJ:b2fpk2m) ([raw file](https://github.com/mirantis/virtlet/blob/963f4c07801fb2672488f11f3460aba601c5e11f/cmd/flexvolume_driver/flexvolume_driver.go#L148)):*
> ```Go
> 
> type cmdInfo struct {
> 	argsNum     int
> ```

numArgs

---

*[tests/e2e/run_ceph.sh, line 44 at r3](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhMz_OfNB_MUbRA7IRX:-KhMz_OfNB_MUbRA7IRY:b-91i64m) ([raw file](https://github.com/mirantis/virtlet/blob/963f4c07801fb2672488f11f3460aba601c5e11f/tests/e2e/run_ceph.sh#L44)):*
> ```Shell
>   fi
>   sleep 2
>   ((ntries = ntries - 1))
> ```

`((ntries--))`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239#-:-KhMzgnrFIpf2aFdl42_:b6r84zn)*
<!-- Sent from Reviewable.io -->

vefimova commented:
Fixed comments and have tests passed. If no objections, let's merge it?

---

Review status: 2 of 14 files reviewed at latest revision, 14 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239#-:-KhWebaVOpq1KseBFrED:b-2b39do)*
<!-- Sent from Reviewable.io -->

jellonek commented:
Reviewed 10 of 13 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[cmd/flexvolume_driver/flexvolume_driver.go, line 116 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhWg9dbGStOvtmGHqjW:-KhWg9dbGStOvtmGHqjX:buhon96) ([raw file](https://github.com/mirantis/virtlet/blob/9af3f6b395cf97d9e8ea1a8f8aca5bce7ddd17ae/cmd/flexvolume_driver/flexvolume_driver.go#L116)):*
> ```Go
> func Mount(args []string) {
> 	target := args[0]
> 	// device := args[1]
> ```

Commented code.

---

*[pkg/libvirttools/virtualization.go, line 380 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhWgr8hQGABlLFwgYHC:-KhWgr8hQGABlLFwgYHD:bwyfhyv) ([raw file](https://github.com/mirantis/virtlet/blob/9af3f6b395cf97d9e8ea1a8f8aca5bce7ddd17ae/pkg/libvirttools/virtualization.go#L380)):*
> ```Go
> 	for _, volXML := range volumesXML {
> 		err = addDiskToDomainDefinition(domain, volXML)
> 		if err != nil {
> ```

Not done. This err assign can be in this if.

---

*[tests/e2e/run_ceph.sh, line 51 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhWhMl78zszDj-iXtBi:-KhWhMl78zszDj-iXtBj:b6gre13) ([raw file](https://github.com/mirantis/virtlet/blob/9af3f6b395cf97d9e8ea1a8f8aca5bce7ddd17ae/tests/e2e/run_ceph.sh#L51)):*
> ```Shell
> 
> # Adjust ceph configs
> docker exec ${container_name} /bin/bash -c 'echo "rbd default features = 1\nrbd default format = 2" >> /etc/ceph/ceph.conf'
> ```

Missing `-e`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

vefimova commented:
Review status: 11 of 14 files reviewed at latest revision, 17 unresolved discussions.

---

*[test.sh, line 14 at r3](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhMzvK7lQ73eA44OjMU:-KhWizYdBi0fvLTx9qQi:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/963f4c07801fb2672488f11f3460aba601c5e11f/test.sh#L14)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

hmm maybe just `cd "${SCRIPT_DIR}"` after `SCRIPT_DIR=...` ?
</blockquote></details>

Done.

---

*[cmd/flexvolume_driver/flexvolume_driver.go, line 148 at r3](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhN-DaNqJfXg92UxcJI:-KhWjJEzTb3SXzTurpyL:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/963f4c07801fb2672488f11f3460aba601c5e11f/cmd/flexvolume_driver/flexvolume_driver.go#L148)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

numArgs
</blockquote></details>

Done.

---

*[cmd/flexvolume_driver/flexvolume_driver.go, line 116 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhWg9dbGStOvtmGHqjW:-KhWj-s9NLSkKJNnvGsm:b-bni4g7) ([raw file](https://github.com/mirantis/virtlet/blob/9af3f6b395cf97d9e8ea1a8f8aca5bce7ddd17ae/cmd/flexvolume_driver/flexvolume_driver.go#L116)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Commented code.
</blockquote></details>

This was left by intention as device name in not used, but be aware when looking, what is supposed to be passed in second arg.

---

*[docs/volumes.md, line 159 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh350KRxwOWUP28ph_F:-KhWkAFy8VYDVzwjoJ4p:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/docs/volumes.md#L159)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

This is not clear (at last to someone who don't know internals) how these volumes will be attached and how they will be seen by VM.
They will be provided as disks, but it's up to VM os/configuration what will be done with them.
</blockquote></details>

Done.

---

*[pkg/libvirttools/storage_pool.go, line 285 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh3-dK7nYMI8nUhSPyg:-KhWkCGguTG-0UB3jwKc:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/pkg/libvirttools/storage_pool.go#L285)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Clarified in later comment - already oked ;)
</blockquote></details>

Done.

---

*[pkg/libvirttools/virtualization.go, line 380 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhWgr8hQGABlLFwgYHC:-KhWllOANzmxDy3Os_U6:bkk9vu3) ([raw file](https://github.com/mirantis/virtlet/blob/9af3f6b395cf97d9e8ea1a8f8aca5bce7ddd17ae/pkg/libvirttools/virtualization.go#L380)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Not done. This err assign can be in this if.
</blockquote></details>

yep, missed it, fixed.

---

*[tests/e2e/run_ceph.sh, line 31 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/239#-Kh2_qtzQIGUD6s_gSh9:-KhWkLkZbzOHueWs_48d:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/08812cddfbe62a6c8410d26d5640ecb0c3171b4a/tests/e2e/run_ceph.sh#L31)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

You can use https://github.com/Mirantis/virtlet/blob/master/pkg/utils/wait.go#L25 as in https://github.com/Mirantis/virtlet/blob/d2dfc5b785b6106a50e7b264ac3c96af77b97fa7/tests/integration/container_statuses_test.go#L130
</blockquote></details>

Done.

---

*[tests/e2e/run_ceph.sh, line 44 at r3](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhMz_OfNB_MUbRA7IRX:-KhWkM_KnepuKt7qwfVa:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/963f4c07801fb2672488f11f3460aba601c5e11f/tests/e2e/run_ceph.sh#L44)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

`((ntries--))`
</blockquote></details>

Done.

---

*[tests/e2e/run_ceph.sh, line 51 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhWhMl78zszDj-iXtBi:-KhWkN1TiDoJZ8sVhouk:beaemtz) ([raw file](https://github.com/mirantis/virtlet/blob/9af3f6b395cf97d9e8ea1a8f8aca5bce7ddd17ae/tests/e2e/run_ceph.sh#L51)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Missing `-e`.
</blockquote></details>

yep,thx

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

jellonek commented:
Review status: 11 of 14 files reviewed at latest revision, 17 unresolved discussions.

---

*[cmd/flexvolume_driver/flexvolume_driver.go, line 116 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhWg9dbGStOvtmGHqjW:-KhWnZRE-mn2rbDEGRIw:b-140k7t) ([raw file](https://github.com/mirantis/virtlet/blob/9af3f6b395cf97d9e8ea1a8f8aca5bce7ddd17ae/cmd/flexvolume_driver/flexvolume_driver.go#L116)):*
<details><summary><i>Previously, vefimova (Victoria Efimova) wrote…</i></summary><blockquote>

This was left by intention as device name in not used, but be aware when looking, what is supposed to be passed in second arg.
</blockquote></details>

So instead leaving commented code - func can be documented by docstring, right?

Btw. why all all funcs there are exported/capitalized. They are in `main` package.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

vefimova commented:
Review status: 11 of 14 files reviewed at latest revision, 17 unresolved discussions.

---

*[cmd/flexvolume_driver/flexvolume_driver.go, line 116 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhWg9dbGStOvtmGHqjW:-KhWyfoUiZxFk5lXiB0P:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/9af3f6b395cf97d9e8ea1a8f8aca5bce7ddd17ae/cmd/flexvolume_driver/flexvolume_driver.go#L116)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

So instead leaving commented code - func can be documented by docstring, right?

Btw. why all all funcs there are exported/capitalized. They are in `main` package.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

vefimova commented:
Rebased and fixed e2e failure, please check.

---

Review status: 6 of 14 files reviewed at latest revision, 17 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239#-:-KhXpP_RAg26Zehe00Tr:bi01leo)*
<!-- Sent from Reviewable.io -->

jellonek commented:
Small typo and then <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 1 of 13 files at r3, 8 of 8 files at r5.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[cmd/flexvolume_driver/flexvolume_driver.go, line 30 at r5](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhXr5qfpGb1tZRPZdo4:-KhXr5qfpGb1tZRPZdo5:b-807afv) ([raw file](https://github.com/mirantis/virtlet/blob/af7caa580cd5bf093784d4f2c7a6705d4aa95f67/cmd/flexvolume_driver/flexvolume_driver.go#L30)):*
> ```Go
> 
> type attachOpts struct {
> 	AsType    string `json:"kubernetes.io/fsType"`
> ```

Typo.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239#-:-KhXroPpeS_cJw3d4Fta:b8jc5tk)*
<!-- Sent from Reviewable.io -->

vefimova commented:
Review status: 9 of 14 files reviewed at latest revision, 14 unresolved discussions.

---

*[cmd/flexvolume_driver/flexvolume_driver.go, line 30 at r5](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhXr5qfpGb1tZRPZdo4:-KhXu74zSVKcSmiBy_fz:b78uebk) ([raw file](https://github.com/mirantis/virtlet/blob/af7caa580cd5bf093784d4f2c7a6705d4aa95f67/cmd/flexvolume_driver/flexvolume_driver.go#L30)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Typo.
</blockquote></details>

thx, fixed!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

jellonek commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 13 files at r3, 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239#-:-KhXv4SMkr1YADaehbur:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jellonek commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[cmd/flexvolume_driver/flexvolume_driver.go, line 31 at r6](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhXzQWNjXSH61yQrzTX:-KhXzQWNjXSH61yQrzTY:bjnkcud) ([raw file](https://github.com/mirantis/virtlet/blob/37e1d810de9022cc52aadcbd95959332c159dd94/cmd/flexvolume_driver/flexvolume_driver.go#L31)):*
> ```Go
> type attachOpts struct {
> 	FsType    string `json:"kubernetes.io/fsType"`
> 	Readwrite string `json:"kubernetes.io/readwrite"`
> ```

`FsType` and `Readwrite` are not used there at all. So this is dead code. IMO this should be removed as it's unnecessary part to maintain, which is - i'll repeat - not used.
IMO there is no point to define this, if whole struct is not even once marshaled/used in any error string.
Instead - this provides a code (struct definition is still a code) lines to maintain, and as example with previous typo there show - error prone.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

vefimova commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[cmd/flexvolume_driver/flexvolume_driver.go, line 31 at r6](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhXzQWNjXSH61yQrzTX:-KhY0b0qsGIgM3R001S3:b3d1qxj) ([raw file](https://github.com/mirantis/virtlet/blob/37e1d810de9022cc52aadcbd95959332c159dd94/cmd/flexvolume_driver/flexvolume_driver.go#L31)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

`FsType` and `Readwrite` are not used there at all. So this is dead code. IMO this should be removed as it's unnecessary part to maintain, which is - i'll repeat - not used.
IMO there is no point to define this, if whole struct is not even once marshaled/used in any error string.
Instead - this provides a code (struct definition is still a code) lines to maintain, and as example with previous typo there show - error prone.  
</blockquote></details>

they are used to check API contract thru unmarshaling

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

jellonek commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[cmd/flexvolume_driver/flexvolume_driver.go, line 31 at r6](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhXzQWNjXSH61yQrzTX:-KhY1HtR54FTnhaksMj2:b-wkgx7q) ([raw file](https://github.com/mirantis/virtlet/blob/37e1d810de9022cc52aadcbd95959332c159dd94/cmd/flexvolume_driver/flexvolume_driver.go#L31)):*
<details><summary><i>Previously, vefimova (Victoria Efimova) wrote…</i></summary><blockquote>

they are used to check API contract thru unmarshaling
</blockquote></details>

But what this check gives to us? Failure on unused by us keys?
These lines only adds bloat, which has to be covered by tests.
Also if someone will change names on other side - we will need additional update on our side, even when we do not use theirs values, so this validation of contract would be then harmful, instead being helpful.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

vefimova commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[cmd/flexvolume_driver/flexvolume_driver.go, line 31 at r6](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhXzQWNjXSH61yQrzTX:-KhY4MgBjSpRJF2B862A:bhnloed) ([raw file](https://github.com/mirantis/virtlet/blob/37e1d810de9022cc52aadcbd95959332c159dd94/cmd/flexvolume_driver/flexvolume_driver.go#L31)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

But what this check gives to us? Failure on unused by us keys?
These lines only adds bloat, which has to be covered by tests.
Also if someone will change names on other side - we will need additional update on our side, even when we do not use theirs values, so this validation of contract would be then harmful, instead being helpful.
</blockquote></details>

This is how this driver API is designed, they can change any function name, parameter set and passing options, so that will be another hint to revisit API.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

vefimova commented:
Review status: 13 of 14 files reviewed at latest revision, 14 unresolved discussions.

---

*[cmd/flexvolume_driver/flexvolume_driver.go, line 31 at r6](https://reviewable.io:443/reviews/mirantis/virtlet/239#-KhXzQWNjXSH61yQrzTX:-KhaT7swr7Cu8RZcY7wg:bqq3ctf) ([raw file](https://github.com/mirantis/virtlet/blob/37e1d810de9022cc52aadcbd95959332c159dd94/cmd/flexvolume_driver/flexvolume_driver.go#L31)):*
<details><summary><i>Previously, vefimova (Victoria Efimova) wrote…</i></summary><blockquote>

This is how this driver API is designed, they can change any function name, parameter set and passing options, so that will be another hint to revisit API.
</blockquote></details>

ok, removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

jellonek commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 13 files at r3, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239#-:-KhaqbW6XOTWqGapaKjt:bnfp4nl)*
<!-- Sent from Reviewable.io -->

ivan4th commented:
Reviewed 1 of 16 files at r1, 1 of 2 files at r2, 5 of 13 files at r3, 1 of 3 files at r4, 3 of 8 files at r5, 4 of 5 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239)*
<!-- Sent from Reviewable.io -->

ivan4th commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/239#-:-Khau6Tvx9pBFwQw5ag2:bnfp4nl)*
<!-- Sent from Reviewable.io -->
